app/web_framework/cluster_base_handler.py

```diff
@@ -62,6 +66,14 @@ def _handle_request_exception(self, ex):
         self.set_status(status_code)
         self.finish()
 
+    def _get_pagination_params(self):
+        offset = self.get_query_argument('offset', self.DEFAULT_OFFSET, True)
+        limit = self.get_query_argument('limit', self.DEFAULT_LIMIT, True)
+        # No negative offsets or limits are accepted
+        non_negative_offset = max(int(offset), 0)
```

Comment:
1. This is sort've a stupid edge case, but `offset` and `limit` are user provided strings, which means `int` might raise a value error and return a 500 to the user. Catch the error and return a 400 instead.

2. Might as well just rename `non_negative_offset` and `non_negative_limit` to `offset` and `limit`

3. If you go with my earlier suggestion (shamelessly stolen from Joey) about allowing `offset` and `limit` to be negative, you'll want to get rid of `max` logic from this method (allow negative numbers to be passed through to the `builds` method.